File: questions and attempted answers/204-count-primes/second-try-with-assistence/204.计数质数.java
/*
 * @lc app=leetcode.cn id=204 lang=java
 *
 * [204] 计数质数
 */

// @lc code=start
class Solution {
    public int countPrimes(int n) {
        int cnt = 0;
        int[] axis = new int[n+1];
        // 这里数轴加一个，是为了避免后续代码index out of bound错误
        Arrays.fill(axis, 1);
        for(int i=2; i<n; i++){
            if (axis[i] == 1){
                cnt++;
                for (int j=i; j<=n/i; j++){
        // java的整数除法，得到的结果省去小数部分，所以要加上等于。
        // 同时考虑到整除的情况，i*j最大是n，给axis增长一个避免报错。
        // leetcode上的代码同时用了if判断和for循环语句当中的判断。
        // 觉得有些多余所以采用了这种写法。
                    axis[i*j] = 0;
                }
            }
        }
        return cnt;
    }
}
// @lc code=end

File: questions and attempted answers/204-count-primes/first-try/204.计数质数.py
#
# @lc app=leetcode.cn id=204 lang=python3
#
# [204] 计数质数
#

# @lc code=start
import random
# from tqdm import tqdm

class Solution:
    def countPrimes(self, n: int) -> int:
        if n < 2:
            return 0
        else:
            cnt = 0
            for i in range(2, n):
                if self.isPrime(i, 10):
                    cnt += 1
        
            return cnt

    # Python3 program Miller-Rabin primality test
 
    # Utility function to do
    # modular exponentiation.
    # It returns (x^y) % p
    def modular_exponentiation(self, x, y, p):
        
        # Initialize result
        res = 1
        
        # Update x if it is more than or
        # equal to p
        x = x % p
        while (y > 0):
            
            # If y is odd, multiply
            # x with result
            if (y & 1):
                res = (res * x) % p
    
            # y must be even now
            y = y>>1; # y = y/2
            x = (x * x) % p
        
        return res
 
    # This function is called
    # for all k trials. It returns
    # false if n is composite and
    # returns false if n is
    # probably prime. d is an odd
    # number such that d*2<sup>r</sup> = n-1
    # for some r >= 1
    def miillerTest(self, d, n):
        
        # Pick a random number in [2..n-2]
        # Corner cases make sure that n > 4
        a = 2 + random.randint(1, n - 4);
    
        # Compute a^d % n
        x = self.modular_exponentiation(a, d, n);
    
        if (x == 1 or x == n - 1):
            return True
    
        # Keep squaring x while one
        # of the following doesn't
        # happen
        # (i) d does not reach n-1
        # (ii) (x^2) % n is not 1
        # (iii) (x^2) % n is not n-1
        while (d != n - 1):
            x = (x * x) % n
            d *= 2
    
            if (x == 1):
                return False
            if (x == n - 1):
                return True
    
        # Return composite
        return False
    
    # It returns false if n is
    # composite and returns true if n
    # is probably prime. k is an
    # input parameter that determines
    # accuracy level. Higher value of
    # k indicates more accuracy.
    def isPrime(self, n, k):
        
        # Corner cases
        if (n <= 1 or n == 4):
            return False
        if (n <= 3):
            return True
    
        # Find r such that n =
        # 2^d * r + 1 for some r >= 1
        d = n - 1
        while (d % 2 == 0):
            d //= 2
    
        # Iterate given nber of 'k' times
        for i in range(k):
            if (self.miillerTest(d, n) == False):
                return False
    
        return True
# @lc code=end

if __name__ == "__main__":
    s = Solution()
    total = s.countPrimes(5000000)
    with open("total_primes.txt", "w") as fi:
        fi.write(str(total))